consensus/src/transactions.rs

```diff
@@ -1,94 +1,104 @@
 };
 
 use futures::FutureExt;
-use monero_serai::ringct::RctType;
-use monero_serai::transaction::Transaction;
+use monero_serai::{
+    ringct::RctType,
+    transaction::{Input, Timelock, Transaction},
+};
 use rayon::prelude::*;
 use tower::{Service, ServiceExt};
 use tracing::instrument;
 
-use cuprate_helper::asynch::rayon_spawn_async;
-use monero_consensus::{
+use cuprate_consensus_rules::{
     transactions::{
-        check_transaction_contextual, check_transaction_semantic, RingCTError, TransactionError,
-        TxRingMembersInfo,
+        check_decoy_info, check_transaction_contextual, check_transaction_semantic,
+        output_unlocked, TransactionError,
     },
     ConsensusError, HardFork, TxVersion,
 };
+use cuprate_helper::asynch::rayon_spawn_async;
 
 use crate::{
-    batch_verifier::MultiThreadedBatchVerifier, context::ReOrgToken, Database, DatabaseRequest,
-    DatabaseResponse, ExtendedConsensusError,
+    batch_verifier::MultiThreadedBatchVerifier,
+    transactions::contextual_data::{batch_get_decoy_info, batch_get_ring_member_info},
+    Database, DatabaseRequest, DatabaseResponse, ExtendedConsensusError,
 };
 
 pub mod contextual_data;
-mod output_cache;
-
-pub use output_cache::OutputCache;
-
-pub async fn batch_setup_txs(
-    txs: Vec<(Vec<Transaction>, HardFork)>,
-) -> Result<Vec<Vec<Arc<TransactionVerificationData>>>, ExtendedConsensusError> {
-    let batch_verifier = Arc::new(MultiThreadedBatchVerifier::new(rayon::current_num_threads()));
-
-    // Move out of the async runtime and use rayon to parallelize the serialisation and hashing of the txs.
-    let txs = rayon_spawn_async(move || {
-        let txs = txs
-            .into_par_iter()
-            .map(|(txs, hf)| {
-                txs.into_par_iter()
-                    .map(|tx| {
-                        Ok(Arc::new(TransactionVerificationData::new(
-                            tx,
-                            &hf,
-                            batch_verifier.clone(),
-                        )?))
-                    })
-                    .collect::<Result<Vec<_>, ConsensusError>>()
-            })
-            .collect::<Result<Vec<_>, ConsensusError>>()?;
-
-        if !Arc::into_inner(batch_verifier).unwrap().verify() {
-            Err(ConsensusError::Transaction(TransactionError::RingCTError(
-                RingCTError::BulletproofsRangeInvalid,
-            )))?
-        }
 
-        Ok::<_, ConsensusError>(txs)
-    })
-    .await?;
+/// A struct representing the type of validation that needs to be completed for this transaction.
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+enum VerificationNeeded {
+    /// Both semantic validation and contextual validation are needed.
+    SemanticAndContextual,
+    /// Only contextual validation is needed.
+    Contextual,
+}
 
-    Ok(txs)
+/// Represents if a transaction has been fully validated and under what conditions
+/// the transaction is valid in the future.
+#[derive(Debug, Clone)]
```

Comment:
Could be `Ord` if `HardFork` did the same.
```suggestion
#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
```

Reply:
I'm not going to make this Ord as I don't think it really makes sense